File: mobile/src/sync/baseline.ts
import {executeSql, initDB} from '../database/index';
import type {FormaSyncManifest} from './manifest';
import {getLastSeenRevision, getLastUploadAt} from './syncMeta';

export async function localHasSyncableData(): Promise<boolean> {
  await initDB();
  const tables = [
    'food_entries',
    'body_metrics',
    'workouts',
    'stretching_log',
    'bracelet_calories_queue',
    'cardio_workouts_cache',
    'food_products_local',
    'strength_presets_cache',
  ];
  for (const table of tables) {
    const rs = await executeSql(
      `SELECT 1 FROM ${table} WHERE (deleted_at IS NULL OR deleted_at = '') LIMIT 1`,
    );
    if (rs.rows.length > 0) {
      return true;
    }
  }
  const prefs = await executeSql('SELECT 1 FROM forma_sync_preferences LIMIT 1');
  return prefs.rows.length > 0;
}

export async function needsBaselineUpload(
  remoteManifest: FormaSyncManifest | null,
): Promise<boolean> {
  if (remoteManifest != null) {
    return false;
  }
  const hasData = await localHasSyncableData();
  if (!hasData) {
    return false;
  }
  const localRevision = await getLastSeenRevision();
  if (localRevision > 0) {
    return false;
  }
  const lastUpload = await getLastUploadAt();
  if (lastUpload) {
    return false;
  }
  return true;
}

File: backend/services/forma_sync/baseline.py
# -*- coding: utf-8 -*-
from __future__ import annotations

import sqlite3

from backend.database.db_utils import get_current_user_id
from backend.services.forma_sync.manifest import FormaSyncManifest

ACTIVE_NOT_DELETED = "(deleted_at IS NULL OR deleted_at = '')"


def local_has_syncable_data(conn: sqlite3.Connection, user_id: int | None = None) -> bool:
    uid = user_id if user_id is not None else get_current_user_id()
    tables_with_user = (
        "food_entries",
        "stretching_log",
        "cardio_workouts",
        "workout_presets",
    )
    for table in tables_with_user:
        try:
            row = conn.execute(
                f"SELECT 1 FROM {table} WHERE user_id = ? AND {ACTIVE_NOT_DELETED} LIMIT 1",
                (uid,),
            ).fetchone()
            if row:
                return True
        except sqlite3.OperationalError:
            continue
    for table in ("body_metrics", "daily_bracelet_calories"):
        try:
            row = conn.execute(
                f"SELECT 1 FROM {table} WHERE user_id = ? AND {ACTIVE_NOT_DELETED} LIMIT 1",
                (uid,),
            ).fetchone()
            if row:
                return True
        except sqlite3.OperationalError:
            continue
    try:
        row = conn.execute(
            f"""SELECT 1 FROM strength_workouts
                WHERE user_id = ? AND {ACTIVE_NOT_DELETED} LIMIT 1""",
            (uid,),
        ).fetchone()
        if row:
            return True
    except sqlite3.OperationalError:
        pass
    try:
        row = conn.execute("SELECT 1 FROM forma_sync_touch LIMIT 1").fetchone()
        if row:
            return True
    except sqlite3.OperationalError:
        pass
    return False


def needs_baseline_upload(
    remote_manifest: FormaSyncManifest | None,
    local_revision: int,
    last_upload_at: str | None,
    has_local_data: bool,
) -> bool:
    if remote_manifest is not None:
        return False
    if not has_local_data:
        return False
    if local_revision > 0:
        return False
    if last_upload_at:
        return False
    return True
